### DISCUSSION TITLE
Upload Cesium 3d File using Reality Data API

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#3141] [Q&A] [answered]
I am trying to upload cesium3d file using Reality data API. But I am getting Authentication Error in the upload step.

1. I am first creating a reality data, using following values.
"realityData": {
        "displayName": "SampleModel",
        "classification": "Model",
        "type": "Cesium3DTiles",
        "description": "Description of the reality data",
        "dataset": "Dataset",
        "group": "GroupId",
        "rootDocument": "Models/Scene/myModel.3mx",
...
}

2. Then i make a upload request using postman following the reality data api tutorial, using this request https://developer.bentley.com/tutorials/realitydata-api-quick-start/#31-upload-a-reality-model

(in this step i am not sure if i should add the modelfile in the body as this is a GET request, so correct me if i am wrong).

3. I receive a SASurl as a response to the above request. Then I make a PUT request on the SASurl, with x-ms-blob-type: BlockBlob as a header. And i get the following response,

![image](https://user-images.githubusercontent.com/89520756/151933345-0d4f7d92-e2e8-4322-a8a1-78e92126892d.png)

4. I tried adding the Authorization header with "bearer tokenString" as its value. but then i get following response, but then i get the following response,

![image](https://user-images.githubusercontent.com/89520756/151933787-30746ac6-06a2-40c2-833b-d965c5aff339.png)

kindly let me know if i am missing something or doing anything wrong.
Thank You.

**Comment by hanan-commits-work** [comment, marked as answer]
Any update on this?

**Reply by EricPaquetB** [reply]
Hello (and sorry for the late reply), 
First let me answer a sub-question: 
"in this step i am not sure if i should add the modelfile in the body..." : no you don't need a body here, and I recommend you to take a look at the APIs documentation on https://developer.bentley.com/, particularly the endpoint that you're using: https://developer.bentley.com/apis/reality-data/operations/get-reality-data-container/; you'll get the exact usage of the APIs.

Now, I'm not sure why you get an AuthenticationFailed on your step 3, but you don't need for sure an Authorization header, as you suggest in step 4.

We'll continue to investigate to reproduce exactly your issue. But in the meantime, if the instructions above unblock you, let us know.

Thanks,
Eric

**Reply by EricPaquetB** [reply]
Hello, 
I think I understand your issue. Probably that you're trying to use the SAS URL directly in Postman. It's not the way it's meant to be used.

Once you get the SAS, you have to use the Azure Storage API to read/write/navigate/... the container. For example, here is a sample TypeScript function that uploads some text to a blob. Here, the "realityDataContainerUrl" parameter is your SAS URL.

```
    private async uploadBlob_realityData(realityDataContainerUrl: string)
    {
        const blobContent = "This is a sample .3mx file";
        const blobName = "samples/sample.3mx";
        const blockBlobClient =  new ContainerClient(realityDataContainerUrl).getBlockBlobClient(blobName);
        const uploadBlobResponse = await blockBlobClient.upload(blobContent, Buffer.byteLength(blobContent));

        console.log(`Uploaded ${blobName} successfully`, uploadBlobResponse.requestId);
    }
```

This function uses the ContainerClient class (part of Azure SDK) to write some content to the container. You can find the documentation for this class here: https://docs.microsoft.com/en-us/javascript/api/@azure/storage-blob/containerclient?view=azure-node-latest#@azure-storage-blob-containerclient-constructor-2 . This is the JavaScript flavor.

Now, I realize that our documentation is not very clear and may lead to the confusion that you experienced. We take good note of that and we'll try to improve our instructions.

HTH,
Eric

**Reply by hanan-commits-work** [reply]
Thank you. I will look into it.

**Reply by hanan-commits-work** [reply]
Kindly clarify one more thing...
Is it essential to have an Azure storage account to use sas url? Or I could use it without any account (considering it is a shared access)?

**Reply by EricPaquetB** [reply]
No, you don't need an Azure storage account to use the SAS URL. The SAS gives you access for some time (24 hr if I remember correctly); when it is expired, you have to request another one.

**Reply by hanan-commits-work** [reply]
I created the script as you as you recommended but for uploading the file. The file is being correctly uploaded in connect-ContextShare, but when i try to view it, I get the following error.
![image](https://user-images.githubusercontent.com/89520756/153183037-c3db04d5-0b47-43e9-99a5-f638c7a9edff.png)

Let me walk you through what I am doing step by step, so you can point out any mistake along the way.

1. I create the Reality Model using Reality Data API, providing my project id, sample file name, and Type. This is the response I am getting, 
{
    "realityData": {
        "id": "38a13422-499e-4305-8021-413b61e32443",
        "displayName": "Sample Dragon 4",
        "dataset": "Dataset",
        "group": "GroupId",
        "description": "Description of the reality data",
        "rootDocument": "Models/Scene/myModel.cesium3dtiles",
        "size": 0,
        "classification": "Model",
        "type": "Cesium3DTiles",
        "authoring": false,
 ......
}

Please note that I am trying to upload Cesium3d file, so the type is Cesium 3D, I am not sure if i should put the type cesium3dtiles in the rootDocument but I did it just to be on a safe side, please let me know if this is incorrect.

2. Then I use postman to make a GET request, as mentioned in the tutorial and I get SASurl as a response.
3. Then I use the SASurl in the script (i will be uploading the script here as well), which takes the .zip file in the folder, and uploads it using SASurl (I have attached the sample reality data in the file as well along with the script).
[ScriptforUpload.zip](https://github.com/iTwin/itwinjs-core/files/8031928/ScriptforUpload.zip)
4. I then look at the contextshare portal after a while, and notice that the size of the file is exactly the size of the file i am uploading. But i get the error when I try to view it.
5. I downloaded the file to make sure if the data is same, and the file was exactly as i uploaded. 

Kindly look at it, and let me know if i am doing anything wrong. My suspicions are that there is something wrong _ am doing in Step 1, or with the format of the file I am uploading.

**Reply by EricPaquetB** [reply]
Hello,
Glad to see that you could successfully upload your data to ContextShare !

I see why you have an issue displaying the model. It is because you uploaded the "zip" file. You must upload all files individually to the Azure container (not zipped), and keep the same directory structure of course.

Also, it's important that the rootDocument points to your json file (in your case, tileset.json). The first thing the viewer does is to look where the root document points and read this file to find out which data format it must display. I'm not sure it's essential, but place tileset.json at the root of your container. So the rootDocument should be "tileset.json". 

Here again, I realize that we're lacking some important instructions in our documentation. We'll fix that.

-Eric

**Reply by hanan-commits-work** [reply]
Thank You for the clarification Eric.
I am now uploading all the files individually, and gave the rootDocument = tileset.json. But I am still getting the same error when I try to open the Reality Model in ContextShare
[Script for Azure.zip](https://github.com/iTwin/itwinjs-core/files/8038945/Script.for.Azure.zip)
maybe something is wrong with the script or something...

**Reply by EricPaquetB** [reply]
Hello,
I reworked your script a bit:
[azureTest.zip](https://github.com/iTwin/itwinjs-core/files/8044025/azureTest.zip)

It did not work because you were trying to create a container for each file. So, to clarify the situation, a reality data is always related to one and only one container (which may contain many blobs). Each file is uploaded to the container in its own blob. 

The Azure Storage SDK is a bit complicated, so it requires some time to get used to it. But there is good documentation from Microsoft if you need to do more sophisticated stuff with the SDK.

And I almost forgot. When you create your Reality Data, some fields must be set correctly, particularly your rootDocument and the type (must == 'Cesium3DTiles'). Here are the properties of the reality data that I created to test with your data:

{
  "realityData": {
    "id": "f2d45f79-5e50-48ac-bef5-000000000000",
    "displayName": "DragonCesium3d",
    "description": "Test data from hannan",
    "rootDocument": "tileset.json",
    "size": 0,
    "classification": "Model",
    "type": "Cesium3DTiles",
    "accessControl": "Project",
    "dataCenterLocation": "North Europe",
    "modifiedDateTime": "2022-02-10T20:29:55Z",
    "lastAccessedDateTime": "2022-02-10T20:29:55Z",
    "createdDateTime": "2022-02-10T20:29:55Z"
  }
}

So I ran azureTest.js within a Node app and the files were correctly created. And I could see the result model in our viewer (a nice dragon).

HTH,
Eric

**Reply by hanan-commits-work** [reply]
Thank you so much Eric. You are a lifesaver!

**Reply by EricPaquetB** [reply]
Happy to help you ! Good luck with your project !
